### ISSUE TITLE
diptables does not enforce rules if they are not saved

### ISSUE THREAD

**Issue by river-fall** [wk8/cookbook-diptables#3]
Hi,

If you append a new rule via console "iptables -A -p tcp -j ACCEPT" without saving it in /etc/sysconfig/iptables, rule will work but diptables cookbook won't fix it.

Enforcing saved rules works fine (added diptables recipe to run-list)
CentOS 6.5

**Comment by wk8** [comment]
Hi @river-fall ,

That's intended behavior, I'm afraid. From the README (https://github.com/wk8/cookbook-diptables/blob/1.0.0/README.md#usage) : 
```
This cookbook rebuilds the entire set of rules at every Chef-client run, thus making it trivial to keep the iptables configuration up-to-date.
```
Any rule _not_ defined using diptables resources will be removed when the cookbook runs. In particular any rule you might have defined manually.
Besides, defining rules manually kinda defeats the purpose of using Chef in the first place?

Happy to answer any other question you might have :-)

**Comment by river-fall** [comment]
Hi wk8,
I'm afraid you got me wrong :)

If you manually add any rule via bash console on the node (not defined using diptables), the chef-client won't change the configuration back to desired in the cookbook.

In other words, this cookbook looks for changes only in /etc/sysconfig/iptables file, not for rules running in RAM.

**Comment by wk8** [comment]
@river-fall : I got you wrong indeed, apologies for that :-)

That being said, if I understand you correctly this time: you converge the node once, then manually add a rule, then converge it again, and the rule you added manually is still there. Is that correct?

If it is, that's also intended (albeit arguably more debatable): this cookbook runs on the assumption that nothing else than itself is gonna mess with iptables on your nodes, so if the file containing the rules (/etc/sysconfig/iptables on CentOS) it has generated has not been modified since the last run, it won't bother making iptables reload it.
If, on the other hand, you add or remove another (totally unrelated) rule from your Chef code, you'll see that your manually added rule will indeed disappear when Chef runs again.

If that's an issue for you, I guess we could force re-loading even the file has not been modified, via a node attribute. However, I'm not sure why you would want that? Would you mind elaborating a bit why this is an issue for you?

**Comment by river-fall** [comment]
It could be an issue in that case if someone else (e.g., developer team) got sudo access to servers besides yourself and the chef-server: they could add some rules which you won't notice and they will be kept until reboot/iptables reconfig.

Of course, it is not a major issue, but if you add such kind of RAM check in the next version, it could be handy.

Many thanks!

**Comment by wk8** [comment]
Fair enough. I'll try to add support for that when I have some time, hopefully soon. Would a `['diptables']['force-reload']` node attribute that would make it reload the config regardless of whether it has been modified suit your needs?

**Comment by river-fall** [comment]
I'm afraid it won't help: as soon as I can add 'execute "/etc/init.d/iptables restart"' manually in any reciept so it will restart iptables every 1 hour :)

It would be great if iptables restarted only if there are changes in file/RAM; of course, it is more complicated task, I don't expect it too soon.
Probably I'll try to add it by myself )

Thanks again!

**Comment by wk8** [comment]
@river-fall : I'll keep it in mind, actually not such a complicated thing to do :-)

Thanks for the feedback, always nice to know people find this stuff useful!

(quite unrelated, but you might want to just do iptables reload instead of restart)

**Comment by river-fall** [comment]
Yes, your cookbook is really good and helpful! It is far superior than others due to automatic deletion.

Minor update:
On CentOS 6.x it shows message in log:
"[2015-05-26T04:20:03-04:00] WARN: Don't know how to set up automatic iptables on your distribution, sorry. Please submit a bug ticket at https://github.com/wk8/cookbook-iptables/issues"

But everything works fine.

**Comment by wk8** [comment]
Fixed in version 1.0.1: https://github.com/wk8/cookbook-diptables/commit/873194b4bd1483e88174bb85302929ed07083bca
Please read the comments in that patch though ;-)
